src/Adapter/MSTest.TestAdapter/Discovery/AssemblyEnumerator.cs

```diff
@@ -227,25 +237,23 @@ private List<UnitTestElement> DiscoverTestsInType(
         try
         {
             typeFullName = type.FullName;
-            TypeEnumerator testTypeEnumerator = GetTypeEnumerator(type, assemblyFileName, discoverInternals, discoveryOption, testIdGenerationStrategy);
+            TypeEnumerator testTypeEnumerator = GetTypeEnumerator(type, assemblyFileName, discoverInternals, testIdGenerationStrategy);
             ICollection<UnitTestElement>? unitTestCases = testTypeEnumerator.Enumerate(out ICollection<string> warningsFromTypeEnumerator);
             warningMessages.AddRange(warningsFromTypeEnumerator);
 
             if (unitTestCases != null)
             {
                 foreach (UnitTestElement test in unitTestCases)
                 {
-                    if (discoveryOption == TestDataSourceDiscoveryOption.DuringDiscovery)
+                    if (_typeCache.GetTestMethodInfoForDiscovery(test.TestMethod) is { } testMethodInfo)
                     {
-                        Lazy<TestMethodInfo?> testMethodInfo = GetTestMethodInfo(sourceLevelParameters, test);
-
                         // Add fixture tests like AssemblyInitialize, AssemblyCleanup, ClassInitialize, ClassCleanup.
-                        if (MSTestSettings.CurrentSettings.ConsiderFixturesAsSpecialTests && testMethodInfo.Value is not null)
+                        if (MSTestSettings.CurrentSettings.ConsiderFixturesAsSpecialTests)
                         {
-                            AddFixtureTests(testMethodInfo.Value, tests, fixturesTests);
+                            AddFixtureTests(testMethodInfo, tests, fixturesTests);
                         }
 
-                        if (DynamicDataAttached(test, testMethodInfo, tests))
+                        if (TryUnfoldITestDataSources(test, testMethodInfo, dataSourcesUnfoldingStrategy, tests))
```

Comment:
Here, if `TryUnfoldITestDataSources` returned false, we will add the "folded"/"unexpanded" test to the `tests` list. But in `TryUnfoldITestDataSource`, one data source could have been unfolded already.

Reply:
Fixed